### PR TITLE
QALI-465 Update the GET method in campaign controller

### PR DESCRIPTION
**Pull request by joval2**

### CODE REVIEW

File: head/src/main/kotlin/io/qalipsis/core/head/inmemory/StandaloneCampaignReportStateKeeperImpl.kt [hunk 1]
```diff
@@ -186,10 +186,10 @@ internal class StandaloneCampaignReportStateKeeperImpl(
     }
 
     @LogInputAndOutput
-    override suspend fun retrieveCampaignReport(tenant: String, campaignKey: CampaignKey): CampaignExecutionDetails {
+    override suspend fun retrieveCampaignReport(tenant: String, campaignKeys: List<CampaignKey>): Collection<CampaignExecutionDetails> {
```
Comment (ericjesse) [inline head/src/main/kotlin/io/qalipsis/core/head/inmemory/StandaloneCampaignReportStateKeeperImpl.kt:189]:
```suggestion
    override suspend fun retrieveCampaignReport(tenant: String, campaignKeys: Collection<CampaignKey>): Collection<CampaignExecutionDetails> {
```

File: head/src/main/kotlin/io/qalipsis/core/head/inmemory/StandaloneCampaignReportStateKeeperImpl.kt [hunk 2]
```diff
@@ -186,10 +186,10 @@ internal class StandaloneCampaignReportStateKeeperImpl(
...
         join()
-        return campaignStates[campaignKey]?.map { (_, runningScenarioCampaign) ->
-            runningScenarioCampaign.toReport(campaignKey)
-        }?.toCampaignExecutionDetails() ?: throw IllegalArgumentException("No report found for the expected campaign")
+        return campaignKeys.map { campaignStates[it]?.map { (_, runningScenarioCampaign) ->
```
Comment (ericjesse) [inline head/src/main/kotlin/io/qalipsis/core/head/inmemory/StandaloneCampaignReportStateKeeperImpl.kt:191]:
```suggestion
        return campaignKeys.map { campaignStates[it]?.mapNotNull { (_, runningScenarioCampaign) ->
```

File: head/src/main/kotlin/io/qalipsis/core/head/inmemory/StandaloneCampaignReportStateKeeperImpl.kt [hunk 3]
```diff
@@ -186,10 +186,10 @@ internal class StandaloneCampaignReportStateKeeperImpl(
...
+            runningScenarioCampaign.toReport(it)
+        }?.toCampaignExecutionDetails() ?: throw IllegalArgumentException("No report found for the expected campaign") }
```
Comment (ericjesse) [inline head/src/main/kotlin/io/qalipsis/core/head/inmemory/StandaloneCampaignReportStateKeeperImpl.kt:193]:
```suggestion
        }?.toCampaignExecutionDetails() }
```

File: head/src/main/kotlin/io/qalipsis/core/head/jdbc/repository/CampaignRepository.kt [hunk 4]
```diff
@@ -63,6 +63,20 @@ internal interface CampaignRepository : CoroutineCrudRepository<CampaignEntity,
...
+     * Retrieve a collection of campaign in a tenant based on their keys.
+     *
+     * @param tenant the tenant reference
+     * @param campaignKeys the set of campaign keys
+     */
+    @Query(
+        """SELECT *
+            FROM campaign
+            WHERE key IN (:campaignKeys) AND EXISTS 
+            (SELECT * FROM tenant WHERE reference = :tenant AND id = campaign.tenant_id)"""
```
Comment (ericjesse) [inline head/src/main/kotlin/io/qalipsis/core/head/jdbc/repository/CampaignRepository.kt:76]:
```suggestion
            (SELECT 1 FROM tenant WHERE reference = :tenant AND id = campaign.tenant_id)"""
```

File: head/src/main/kotlin/io/qalipsis/core/head/report/CampaignReportProvider.kt [hunk 5]
```diff
@@ -29,10 +29,10 @@ import io.qalipsis.core.head.model.CampaignExecutionDetails
...
     /**
-     * Retrieves the CampaignReport matching the specified [campaignKey].
+     * Retrieves the CampaignReport matching the specified list of [campaignKeys].
      *
      * @param tenant the reference of the tenant owning the data.
-     * @param campaignKey the ID of the campaign to retrieve
+     * @param campaignKeys the list of keys of the campaigns to retrieve
      */
-    suspend fun retrieveCampaignReport(tenant: String, campaignKey: CampaignKey): CampaignExecutionDetails
+    suspend fun retrieveCampaignReport(tenant: String, campaignKeys: List<CampaignKey>): Collection<CampaignExecutionDetails>
```
Comment (ericjesse) [inline head/src/main/kotlin/io/qalipsis/core/head/report/CampaignReportProvider.kt:37]:
```suggestion
    suspend fun retrieveCampaignsReports(tenant: String, campaignKeys: Collection<CampaignKey>): Collection<CampaignExecutionDetails>
```

File: head/src/main/kotlin/io/qalipsis/core/head/report/DatabaseCampaignReportProvider.kt [hunk 6]
```diff
@@ -52,35 +52,41 @@ internal class DatabaseCampaignReportProvider(
...
-            completedMinions = reportEntity?.completedMinions,
-            successfulExecutions = reportEntity?.successfulExecutions,
-            failedExecutions = reportEntity?.failedExecutions,
-            scenariosReports = reportEntity?.scenariosReports?.let { mapScenarioReport(it) }
-                ?: ongoingScenariosDetails(campaignEntity.id),
-        )
+
+    override suspend fun retrieveCampaignReport(
+        tenant: String,
+        campaignKeys: List<CampaignKey>
```
Comment (ericjesse) [inline head/src/main/kotlin/io/qalipsis/core/head/report/DatabaseCampaignReportProvider.kt:58]:
```suggestion
        campaignKeys: Collection<CampaignKey>
```

File: head/src/main/kotlin/io/qalipsis/core/head/report/DatabaseCampaignReportProvider.kt [hunk 7]
```diff
@@ -52,35 +52,41 @@ internal class DatabaseCampaignReportProvider(
...
-            zones = campaign.zones,
-            startedMinions = reportEntity?.startedMinions,
-            completedMinions = reportEntity?.completedMinions,
-            successfulExecutions = reportEntity?.successfulExecutions,
-            failedExecutions = reportEntity?.failedExecutions,
-            scenariosReports = reportEntity?.scenariosReports?.let { mapScenarioReport(it) }
-                ?: ongoingScenariosDetails(campaignEntity.id),
-        )
+
+    override suspend fun retrieveCampaignReport(
```
Comment (ericjesse) [inline head/src/main/kotlin/io/qalipsis/core/head/report/DatabaseCampaignReportProvider.kt:56]:
```suggestion
    override suspend fun retrieveCampaignsReports(
```

File: head/src/test/kotlin/io/qalipsis/core/head/jdbc/repository/CampaignRepositoryIntegrationTest.kt [hunk 8]
```diff
@@ -1467,7 +1467,7 @@ internal class CampaignRepositoryIntegrationTest : PostgresqlTemplateTest() {
     }
 
     @Test
-    fun `should retrieve valid campaign details by their tenant and matching campaign keys in`() =
+    internal fun `should retrieve valid campaign details by their tenant and matching campaign keys in`() =
```
Comment (ericjesse) [inline head/src/test/kotlin/io/qalipsis/core/head/jdbc/repository/CampaignRepositoryIntegrationTest.kt:1470]: Put internal on the class, not on each function.

**Review by ericjesse — CHANGES_REQUESTED** [review]

File: head/src/main/kotlin/io/qalipsis/core/head/inmemory/StandaloneCampaignReportStateKeeperImpl.kt (hunk 3 above)
Comment (joval2) [reply head/src/main/kotlin/io/qalipsis/core/head/inmemory/StandaloneCampaignReportStateKeeperImpl.kt:193]: What to do when the campaignStates map doesn't contain one of the campaignKeys ? I think it's necessary to throw the exception.

Comment (ericjesse) [reply head/src/main/kotlin/io/qalipsis/core/head/inmemory/StandaloneCampaignReportStateKeeperImpl.kt:193]: no exception, just ignore when the key does not exists and only return the values that exist. However, if none can be found, I would return an IllegalArgumentException

**Review by ericjesse — APPROVED** [review]
Fix my comment, and add test on the cases for missing campaign keys if not done yet.

**Review by ericjesse — APPROVED** [review]

**Pull request merged**